README.md

```diff
@@ -1,3 +1,28 @@
 # thursday
 
 Classifying radio galaxies as Fanaroff-Riley Type I (FR-I) and Fanaroff-Riley Type II (FR-II).
+
+
+## Input files
+
+You will need
+- The data used in https://arxiv.org/pdf/1705.03413.pdf. This is stored in `asu.tsv`.
+
+
+## Running the code
+
+1. Run `fri-frii-download.ipynb` to get all FRI/FRII samples in fri-cat, frii-cat, and FIRST. Data will be stored in as `data.h5` in (current_directory/data/data.h5)
+2. Use the get_data function in `format_data.py` to generate the training and testing data (train_x, train, y, test_x, test_y) from `data.h5` and `asu.tsv`.
+3. Use the `data_gen` function in `format_data.py` to create the data generator.
+4. Instantiate `HognetModel` class in `model_hognet.py`. Do the following with `HognetModel` class methods: Use `build` method to construct the model, and the `train method` to train the model, and then the `load` method to return the fully trained model (This has not yet been pushed due to some highly irritating tensorflow issues).
+5. Instantiate desired `SklearnModel` class in `models.py`. Use `train` method to build and train the sklearn classifier, and then the `load` method to return the fully trained model. If comparing a sklearn model to hognet, use sklearn after hognet (keras uses a generator and stops training when loss plateaus, and thus the number of epochs for which the data is generated is unknown)
```

Comment:
Do I have to instantiate `HognetModel`? What if I only want to train random forests?

Reply:
ok

Reply:
How should I say this?

Reply:
Ideally the methods would be the same. Can you make them have the same API? And then put both into `models.py`, and just say to initialise whatever model you want.

Reply:
One option is to make a function that generates models from sklearn classes.

```python
def _generate_model(Model):
  model = SklearnModel(Model)
  return model

random_forest = _generate_model(RandomForestClassifier)

def hognet(x, y, z):
  ...
```

Reply:
I don't understand why you would want to do this. That is already what I have, essentially. Replying to the class comment.

Reply:
I maintain I need classes to keep everything simple and user friendly

Reply:
but I'm not exactly an expert, so...

Reply:
I can store both SklearnModel and HognetModel in models.py?

Reply:
But there is just no equivalent to passing the classifier object as an argument (as in SklearnModel) in HognetModel. It constructs it from scratch based upon parameters. And I can't use the same class methods since hognet has a lot of required arguments that sklearn does not.

Reply:
Before I respond, note that I haven't seen `models_hognet`, so I don't know what API it has.

Let's think about an abstract API that isn't what you've implemented — just the API itself. I want to have some code that runs the FRI/FRII classification task. Following your readme, I use `format_data` to get some training data. Then I make the data generator. Then I want to train my favourite classifier, so perhaps I can do something like this:

```python
model = models.RandomForest()
model.train(data)

predict(data, model)
```

Now I want to change to using the fancy new HOGNet classifier that I've heard so much about. The change should be simple:

```python
import models

model = models.HOGNet()
model.train(data)

predict(data, model)
```

There's only been one change to my code (swapping random forests for HOGNet). This is a good API.

Looking at your readme, your API is something like this:

```python
import models

model = models.SklearnModel(RandomForestClassifier)
model.train(data)
model = model.load()

predict(data, model)
```

Then I need to change three lines to get HOGNet working.

```python
import model_hognet

model = model_hognet.HognetModel()
model.build()
model.train(data)
model = model.load()

predict(data, model)
```

1. The models should be in the same module (`models`).
2. If `.build()` is always called, then it should be called in `__init__`.
3. Having the exact same API would let me do something like

```python
import models

kwargs = {'data_path': 'mydata.h5'}
ms = [models.RandomForest, models.HOGNet]
ms = [M(**kwargs).train(data) for M in ms]
```

...or something.

Reply:
Have you seen the test.ipynb? The API is super simple to use - basically identical to sklearn normal  API, bar the load() function, which we need due to very long training times. The difference between hognet is very minimal and can be made identical bar the different parameters for training and an extra Model augment for SklearnModel

Reply:
No, I haven't. It doesn't seem to be pushed?

Reply:
Can you link me to it?

Reply:
Getting the API right is super important for usability later on.

Reply:
Yeah I can do both, as there is no path attribute that is updated. So I can return model from train, while also allowing the .load method to work.

Reply:
Do that then. Once you've fixed the minor issues I've commented on here let me know and I'll merge.